### PR TITLE
envoy support

### PR DESCRIPTION
**Pull request by TaoZou1**
Support using internal VC envoy.
if EnvoyPort != 0, create an url is which envoy specific 
Using ca to verify server cert needs ultra header.

Test Done:
    1. ca file mode:
    ./clean -cluster=domain-c9:d75735a3-2847-45d2-a652-ef2d146afd54 -nsx-user=admin -nsx-passwd='xxx'  -mgr-ip=nsxmanager-ob-22386469-1-dev-integ-nsxt-8791  -envoyhost=localhost -envoyport=1080 -log-level=1 -ca-file=./ca.cert
    2. thumbprint mode:
    ./clean -cluster=domain-c9:d75735a3-2847-45d2-a652-ef2d146afd54 -nsx-user=admin -nsx-passwd='xxx'  -mgr-ip=nsxmanager-ob-22386469-1-dev-integ-nsxt-8791  -envoyhost=localhost -envoyport=1080 -log-level=1 -thumbprint=8bc2fa2b5879c27b1180fa44e5f747832f2ded6be483e3c3d2c4816a38870868

### CODE REVIEW

File: cmd_clean/main.go [hunk 1]
```diff
@@ -53,10 +55,57 @@ func (t *Transport) base() http.RoundTripper {
...
+		transport := &Transport{Base: tr}
+		httpClient = &http.Client{
+			Transport: transport,
+		}
+
+		if cf.EnvoyPort != 0 {
+			//envoy client
+			if len(cf.CaFile) > 0 {
+				extraHeaders := make(map[string]string)
+				extraHeaders["x-vmware-server-tls-cert"] = certPemBytesToHeader(cf.CaFile[0])
```
Comment (wsquan171) [inline cmd_clean/main.go:89]: > cf.CaFile[0]

Do we always use the first NSX manager even with NCP cleanup? I assume this is also something yet to be handled on NCP side of cleanup. Apparently it's also possible for wcp to pass per-manager leaf certs. So we need to bring in the positional thumbprint / cert selection implemented in the normal NCP / operator runtime.

They also mentioned a possibility of mixed use of thumbprint and cert (on a per-mgr basis).

File: cmd_clean/main.go [hunk 2]
```diff
@@ -53,10 +55,57 @@ func (t *Transport) base() http.RoundTripper {
...
+func createHttpClient(cf *config.NSXOperatorConfig, useExternalHttp bool) *http.Client {
+	var httpClient *http.Client = nil
+	if useExternalHttp {
+		tr := &http.Transport{}
+		transport := &Transport{Base: tr}
+		httpClient = &http.Client{
+			Transport: transport,
+		}
+
+		if cf.EnvoyPort != 0 {
```
Comment (wsquan171) [inline cmd_clean/main.go:85]: nit: let's make it consistent that we switch to Envoy mode as long as envoyHost is set, and default to port 1080 if unset.

File: Makefile [hunk 3]
```diff
@@ -63,11 +63,12 @@ test: manifests generate fmt vet envtest ## Run tests.
 .PHONY: build
 build: generate fmt vet ## Build manager binary.
 	go build -o bin/manager cmd/main.go
-
-.PHONY: clean
-clean: generate fmt vet ## Build clean binary.
 	go build -o bin/clean cmd_clean/main.go
 
+.PHONY: clean
```
Comment (zhengxiexie) [inline Makefile:68]: Better commit in another PR, for not envoy related.

File: cmd_clean/main.go (hunk 2 above)
Comment (TaoZou1) [reply cmd_clean/main.go:85]: done

File: cmd_clean/main.go [hunk 4]
```diff
@@ -53,10 +55,57 @@ func (t *Transport) base() http.RoundTripper {
...
+	if err != nil {
+		log.Error(err, "failed to read ca file")
+		return ""
+	}
+	cert := string(certPem)
+	cert = strings.ReplaceAll(cert, "-----BEGIN CERTIFICATE-----", "")
+	cert = strings.ReplaceAll(cert, "-----END CERTIFICATE-----", "")
+	cert = strings.ReplaceAll(cert, "\n", "")
+	return strings.TrimSpace(cert)
+}
```
[the remark this answers is not in the document]
Comment (TaoZou1) [reply cmd_clean/main.go]: update the code. the code in main.go just a demo. the code here is only used for wcp. The certPemBytesToHeader will not be invoked by other modules. so  not put it in the utils.

File: cmd_clean/main.go (hunk 1 above)
Comment (TaoZou1) [reply cmd_clean/main.go]: Now use http client from nsx-operator. so the nsx could be changed if the nsx  is not available.

**Review by wsquan171 — COMMENTED** [review]
looks good overall. one nit

File: pkg/nsx/cluster.go [hunk 5]
```diff
@@ -89,15 +99,74 @@ func NewCluster(config *Config) (*Cluster, error) {
...
+
+		cf := cluster.config
+		if len(cf.CAFile) > 0 {
+			envoyUrl = fmt.Sprintf(EnvoyUrlWithCert, cf.EnvoyHost, cf.EnvoyPort, mgrIP)
+		} else if len(cf.Thumbprint) > 0 {
+			thumbprint := thumbprintToUrlPath(cf.Thumbprint[0])
+			envoyUrl = fmt.Sprintf(
+				EnvoyUrlWithThumbprint, cf.EnvoyHost, cf.EnvoyPort, mgrIP, thumbprint)
+		} else {
+			log.Error(errors.New("no CAFile or Thumbprint provided"), "no CAFile or Thumbprint provided")
```
[the remark this answers is not in the document]
Comment (TaoZou1) [reply pkg/nsx/cluster.go]: remove the log.Error here and add checking in config validate

**Review by wsquan171 — APPROVED** [review]

**Pull request merged**